### DISCUSSION TITLE
How to update private npm dependencies only.

### DISCUSSION THREAD

**Discussion by robg-temp** [renovatebot/renovate#9298] [Request Help]
I've been struggling with this for weeks now without success. Renovate either doesn't update anything, or everything.

**We would like to only update our own npm dependencies, which start with "@acme".**  We use one config as follows (I recently separated the blacklist packageRules and then whitelist):

```
{
  "baseBranches": ["master", "release"]
  "onboarding": true,
  "onboardingConfig": {
    "extends": [ "config:base" ]
  },
  "platform": "github",
  "packageRules": [
    {
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false
    },
    {
      "matchBaseBranches": ["master"],
      "rangeStrategy": "bump",
      "updateTypes": [
        "minor",
        "patch"
      ]
    },
    {
      "matchBaseBranches": ["release"],
      "matchPackagePatterns": ["common"],
      "rangeStrategy": "bump",
      "updateTypes": [
        "patch"
      ]
    }
  ]
}
```
**Does that look correct???**

**Comment by viceice** [comment]
The rules are applied one after another when - match. So the first disable all excerpt the pattern.the next rules are applied on top.

So please describe in more detail with samples what's should happen.

**Reply by robg-temp** [reply]
Thanks so much for the prompt answer!

What I would like to have happen depends on the branch:

On **master** **All** packages starting with **"@acme"** should be updated (minor and patch).
On **release** branches, only the "common" package should be updated (patch only).

**The above config did nothing**, perhaps because the first rule, blacklisted everything, as you mentioned.

During testing I would expect to only see a couple of branches at most because I have incremented the versions of a couple of our own repos - the ones that we want updated.

So I tried removing the first packageRules and putting the target repos as a whitelist:

```
"packageRules": [
    {
      "matchBaseBranches": ["master"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false,
      "rangeStrategy": "bump",
      "updateTypes": [
        "minor",
        "patch"
      ]
    },
    {
      "matchBaseBranches": ["release"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["common-web-assets"],
      "enabled": false,
      "rangeStrategy": "bump",
      "updateTypes": [
        "patch"
      ]
    }
  ]
```

That caused **everything** to be updated:

![image](https://user-images.githubusercontent.com/80468671/112683178-8891a280-8e47-11eb-8344-8cd5092c0126.png)

Here's my package.json file with the desired package to update enclosed in **package name**:

```
{
  "name": "dummy",
  "version": "1.4.0",
  "description": "dummy package",
  "repository": {
    "type": "git",
    "url": "https://github.com/Recognia/dummy.git"
  },
  "dependencies": {
    **"@acme/common": "^1.0.202103261457",**
    "@types/jasminewd2": "2.0.8",
    "core-js": "3.9.1",
    "cross-var": "1.1.0",
    "zone.js": "0.11.4"
  },
  "devDependencies": {
  },
  "engines": {
    "node": ">= 8.0.0",
    "npm": ">= 5"
  },
  "renovate": {}
}
```

**Reply by viceice** [reply]
You only disable minor ond patch for master and patch for release. So all other updates are enabled by default

**Reply by robg-temp** [reply]
I didn't realize that the updateTypes also needed to be reversed.  For now, I went back to basics and only set the packages:

```
"packageRules": [
    {
      "matchBaseBranches": ["master"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false
    },
    {
      "matchBaseBranches": ["release"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["common"],
      "enabled": false
    }
  ]
```

I'll see if that part works at least.

**Reply by robg-temp** [reply]
Something is still wrong.  I simplified the rules as much as I could, but renovate is matching "common" but in master only.

```
{
  "baseBranches":  ["master", "release"],
  "onboarding": true,
  "onboardingConfig": {
    "extends": [ "config:base" ]
  },
  "packageRules": [
    {
      "matchBaseBranches": ["release"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["common"],
      "enabled": false
    }
  ]
}
```

According to the docs, matchBaseBranches should work with excludePackagePatterns.

**Reply by viceice** [reply]
Yes, this rule matches only release branch, so no restrictions on other branches

**Reply by robg-temp** [reply]
I understand that part, but why wasn't "common" updated on the release branch I wonder.  Shouldn't the packageRule target it?

**Reply by robg-temp** [reply]
I updated my rules so that nothing happens in master:

```
"packageRules": [
    {
      "matchBaseBranches": ["Release-test"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["common"],
      "enabled": false
    },
    {
      "matchBaseBranches": ["master"],
      "matchPackagePatterns": ["*"],
      "enabled": false
    }
  ]
```

Looks like this is working...

**Reply by viceice** [reply]
Package rules are excuted in order so first use  disable rules and then other rules

**Reply by robg-temp** [reply]
OK, got it.

I added the update types and that broke it. 

```
"packageRules": [
    {
      "matchBaseBranches": ["Release-test"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["common"],
      "enabled": false,
      "rangeStrategy": "bump",
      "updateTypes": [
        "minor",
        "patch"
      ],
      "automerge": true,
      "automergeType": "branch"
    },
    {
      "matchBaseBranches": ["Sprint-test"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false,
      "rangeStrategy": "bump",
      "updateTypes": [
        "patch"
      ],
      "automerge": true,
      "automergeType": "branch"
    }
  ]
```

If I understand correctly, I have to reverse them, so that if I only want "patch" updates, I'd have to put in every other type? 

```
      "rangeStrategy": "bump",
      "updateTypes": [
        'major','minor',pin,digest,lockFileMaintenance,rollback,bump
      ],
```

**Comment by viceice** [comment]
Easiest solution it to first use your disable rules and use new rules for the other settings.

**Reply by robg-temp** [reply]
Trying that now.  Thanks!

**Reply by robg-temp** [reply]
I creates two separate packageRules: one for branch and package matching, and another for version matching.  Is there a way to combine them to only update minor and patch changes for packages that start with "@acme"?

  ```
   {
      "matchBaseBranches": ["Release-test", "Sprint-test"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "enabled": false
    },
    {
      "separateMajorMinor": true,
      "major": {
        "enabled": false
      },
      "automerge": true,
      "automergeType": "branch"
    }
```
Maybe:

 {
      "matchBaseBranches": ["Release-test", "Sprint-test"],
      "matchPackagePatterns": ["*"],
      "excludePackagePatterns": ["^@acme"],
      "matchUpdateTypes": [
        "major"
      ],
      "enabled": false
    },

 Thanks!!!

**Reply by robg-temp** [reply]
Just wanted to say thanks Michael for all your help.  I don't think I could have gotten this working without you!
